### ISSUE TITLE
[Bug] sync stopped working

### ISSUE THREAD

**Issue by andreicristianpetcu** [mozilla-mobile/fenix#15818]
## Steps to reproduce
Log in with fxa. Try to see synced tabs on other devices.

### Expected behavior
tabs from other devices are visible.

### Actual behavior
![Screenshot_20201010-061511_Firefox_Nightly](https://user-images.githubusercontent.com/922368/95644563-2d9a0800-0ac0-11eb-8b9f-f5113e8cc768.png)
![Screenshot_20201010-061500_Firefox_Nightly](https://user-images.githubusercontent.com/922368/95644564-2ecb3500-0ac0-11eb-8f3e-9fd411fe6668.png)

### Device information

* Android device: one plus 5t
* Fenix version: nightly

**Comment by andreicristianpetcu** [comment]
It initially sayd last sync 6 days ago. I logged out and logged back in again and that is why it says last sync never.
Works fine in beta and stable.

**Comment by rfk** [comment]
Thanks for the report. A couple of observations from poking around in the code related to this:

* The "please enable tab syncing" message correspionds to the `SYNC_ENGINE_UNAVAILABLE` error from the synced-tabs view component.
* However, your choose-what-to-sync view here clearly shows that synced tabs are enabled! So are we somehow firing the wrong error code here?
* I do see a place where we fire `SYNC_ENGINE_UNAVAILABLE` [in response to any error](https://github.com/mozilla-mobile/android-components/blob/1daab7ae885702727228ee8902fcb34bb6ca6085/components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/presenter/DefaultPresenter.kt#L128), could this be masking a different underlying error?
* Could the recent refactor to the FxA state-machine have subtly broken the logic we use for checking which engines are enabled? /cc @grigoryk 

@andreicristianpetcu one thing to try, if you haven't already done so - do an explicit "synchronize now" on both your mobile device and your other device, and see if that makes the tabs appear correctly.

**Comment by andreicristianpetcu** [comment]
2 things:
 -  I kept forcing the sync since I reported this bug. Nightly was completely dead, beta/stable on the phone worked fine
 - today there was some progress: I can send tabs from beta to nightly an from nightly to beta by using the sync now. This is the first time I saw it work since reporting the bug. Nightly still shows last sync 'never' and no synced tabs.

**Comment by andreicristianpetcu** [comment]
Not sure if related https://twitter.com/tublitzed/status/1315845146555166721
![image](https://user-images.githubusercontent.com/922368/95837340-2df90400-0d49-11eb-817c-89b4d9f1952d.png)

Since it's Nightly-only it might not be related but it is worth mentioning

**Comment by grigoryk** [comment]
> However, your choose-what-to-sync view here clearly shows that synced tabs are enabled! So are we somehow firing the wrong error code here?

In the screenshot, the engine selections are greyed out and are disabled, which means that we don't actually know engine state yet (because we haven't synced yet). The UI picks "checked, disabled" as the default unknown state representation. This matches up with "Last synced: never".


@andreicristianpetcu 
> "I can send tabs from beta to nightly an from nightly to beta by using the sync now. This is the first time I saw it work since reporting the bug. Nightly still shows last sync 'never' and no synced tabs."

When you say "i can send tabs from beta to nighty", do you mean Fenix Nightly? So the tabs arrive, but synced tabs aren't displayed?

What about other synced data - history, bookmarks?

The most helpful thing at this point would be seeing the logs. Can you do a full logcat capture (`adb logcat`) as you're starting up Nightly, and then going into Settings->Accounts and trying to sync?

**Comment by grigoryk** [comment]
> I do see a place where we fire SYNC_ENGINE_UNAVAILABLE in response to any error, could this be masking a different underlying error?
> Could the recent refactor to the FxA state-machine have subtly broken the logic we use for checking which engines are enabled? /cc @grigoryk

@rfk both are possible, the masking of errors in the UI is certain. Refactor should not have affected the engine selection, and this is the first time I'm seeing a complaint. But it's possible something broke.

In this case, it does look like we're straight up failing to perform a sync, for some reason. Logs should help clarify what's happening.

**Comment by andreicristianpetcu** [comment]
Here is the logcat result https://my.owndrive.com/index.php/s/XQkj8XaEdXAHq8s

**Comment by grigoryk** [comment]
@andreicristianpetcu thank you, very helpful!

In your logs it all appears to be fine, except that sync isn't actually performed.

It appears that WorkManager tasks aren't running on your device. There are also 1300 or so persisted Glean pings which never got sent (by a WorkManager task that was supposed to send them, I think).

**Comment by grigoryk** [comment]
It appears that WorkManager tasks aren't running specifically on Nightly, since @andreicristianpetcu reports that beta/release both are able to sync, which means WM tasks are able to run (or we wouldn't sync otherwise!).

**Comment by andreicristianpetcu** [comment]
Automatically sync data is enabled in the os.
I uninstalled blokata (dns ad blocker).

**Comment by grigoryk** [comment]
@andreicristianpetcu did uninstalling blokata help in your case? Or sync is still not working for you?

**Comment by andreicristianpetcu** [comment]
@grigoryk it did not help

**Comment by LittleAngelwings** [comment]
Same problem with nightlies on my devices. Huawei P Smart (2018) with Android 9.1.0 and Samsung GalaxyTab A SM-T580 with Android 8.1.0, no Security Apps. The only thing is I have a custom addon list activated in my nightlies.

**Comment by DiegoARL38** [comment]
Same on a Redmi Note 8, MiUI 11.0.9, Android 9.

**Comment by andreicristianpetcu** [comment]
Debug builds work fine. Today's beta works fine which was a nightly when I reported the issue. Here are my about:config changes. I don't remember changing any of them.
![Screenshot_20201022-082522_Firefox_Nightly](https://user-images.githubusercontent.com/922368/96828821-fc66f380-1440-11eb-887f-9f0fdf91e694.png)

**Comment by andreicristianpetcu** [comment]
should I uninstall/reinstall nightly?

**Comment by limuc** [comment]
Doesn't work here too. Tried reinstalling, logging in and out. (But the "Sync now" button does not even work in stable/beta, only sync every 10 minutes, when opened)

**Comment by andreicristianpetcu** [comment]
@limuc you might have a different issue. I only have this in nightly.

**Comment by limuc** [comment]
@andreicristianpetcu Yes seems like it. After reseting everything yesterday once again it seems to work. It's also working in nightly now. I will observe it and open a new bug report if it stops working again.

**Comment by andreicristianpetcu** [comment]
I uninstalled/installed nightly. It works now. Closing.

**Comment by andreicristianpetcu** [comment]
Can it be a dupe of this? https://github.com/mozilla-mobile/fenix/issues/16357

**Comment by aminomancer** [comment]
I'm getting this bug right now. Also, on my desktop (Nightly, currently 88.0a1) the "Mobile Bookmarks" folder randomly disappeared from my places db. The "sync now" button does nothing and all the sync option checkboxes are grayed out. Tried signing out and in and reinstalling. Could just be a coincidence, but I got an email from AMO saying I won't be able to sign extensions unless I enable 2FA. So I enabled 2FA and the very next day is when this bug appeared for me. Never had any problems with syncing or mobile bookmarks until now

**Comment by aminomancer** [comment]
For some reason I was able to fix the problem by re-pairing my phone and manually sending a tab from desktop to android. When I clicked the notification for the sent tab, it opened the link and then abruptly started syncing my mobile tabs. Then when I went to the account settings in the android UI, the sync options were back to normal. Unfortunately, my "Mobile Bookmarks" folder is still missing on desktop for some reason. That coincided with the start of this issue, but apparently not with its resolution. Of course, I can make a folder called "Mobile Bookmarks" manually, but in the inspector I can see that it's not the same thing, and it isn't populated by synced bookmarks. The node for the actual Mobile Bookmarks subview button is still there, #BMB_mobileBookmarks, it's just hidden and has no children. So at the moment it looks like there is no way for me to get that folder back.
